src/ng/directive/input.js

```diff
@@ -1180,6 +1186,26 @@ function baseInputType(scope, element, attr, ctrl, $sniffer, $browser) {
   // or form autocomplete on newer browser, we need "change" event to catch it
   element.on('change', listener);
 
+  // Some native input types (date-family) have the ability to change validity without
+  // firing any input/change events.
+  // For these event types, when native validators are present and the browser supports the type,
+  // check for validity changes on various DOM events.
+  if (PARTIAL_VALIDATION_TYPES[type] && ctrl.$$hasNativeValidators && type === attr.type) {
+    element.on(PARTIAL_VALIDATION_EVENTS, function(ev) {
+      if (!timeout) {
+        var validity = this[VALIDITY_STATE_PROPERTY] || {};
```

Comment:
In `badInputChecker`, we check `element[0].validity` and then call `element.prop(VALIDITY_STATE_PROPERTY)` asynchronousy. I am not sure if it is possible for the node inside `element` to be replaced (e.g. by another directive). If not, then I guess we could change that too.

It's slightly different here, since the listener is attached to the underlying DOM node, so it's guarranteed not to change.